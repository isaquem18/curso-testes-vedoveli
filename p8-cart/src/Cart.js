import find from 'lodash/find';
import remove from 'lodash/remove';

export default class Cart {
  items = [];
  
  // calcular o total a ser pago considerando também condições de descontos
  getTotal() {
    return this.items.reduce((acc, item) => {

      if (item.condition?.percentage) {

        return acc += item.product.price * item.quantity;
      }

      return acc += item.product.price * item.quantity;
    }, 0);
  };

  add(item) {
    const itemToFind = { product: item.product };

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind)
    };

    this.items.push(item);
  };

  remove(product) {
    remove(this.items, { product })
  };

  // retorna tudo que se tem no cart
  summary() {
    const total = this.getTotal();
    const items = this.items;

    return {
      total,
      items
    };
  };

  // retorna tudo que se tem E limpa o cart 
  checkout() {
   const { total, items } = this.summary();
 
    this.items = []

    return {
      total,
      items
    };
  };

}
