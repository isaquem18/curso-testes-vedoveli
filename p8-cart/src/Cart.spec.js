import Cart from "./Cart";


describe("check the Cart class operations", () => {
  let cart;
  let product = {
    title: "Adidas runnign shoes - men",
    price: 35388,
  };

  let product2 = {
    title: "Adidas runnign shoes - women",
    price: 41872,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe("getTotal()", () => {
    it("should return 0 when getTotal() is executed in a newly created instance", () => {
      const cart = new Cart();

      expect(cart.getTotal()).toBe(0);
    });

    it("should multiply quantity and price and receive the total amount", () => {
      const item = {
        product,
        quantity: 2,
      };

      cart.add(item);

      expect(cart.getTotal()).toBe(70776);
    });

    it("should ensure no more than one product exists at a time", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal()).toEqual(35388);
    });

    it("should update total when a product gets included an then removed", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal()).toBe(41872);
    });
  });

  describe("checkout()", () => {
    it("should return an object with the total and the list of items", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it("should reset the cart object when checkout() is called", () => {
      cart.add({
        product,
        quantity: 3,
      });

      cart.checkout();
    });
  });

  describe("summary()", () => {
    it("should return the cart array with the added item", () => {
      cart.add({
        product,
        quantity: 2,
      });

      expect(cart.summary()).toMatchSnapshot();
    });
  });

  describe('special conditions', () => {

    it('should apply percentage discount above minimun is passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2
      };

      cart.add({
        product,
        condition,
        quantity: 2
      });

      expect(cart.getTotal()).toBe(70776);
    });
  });
});
