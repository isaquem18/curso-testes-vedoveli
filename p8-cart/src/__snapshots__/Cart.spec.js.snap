// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the Cart class operations checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas runnign shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas runnign shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`check the Cart class operations summary() should return the cart array with the added item 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas runnign shoes - men",
      },
      "quantity": 2,
    },
  ],
  "total": 70776,
}
`;
